R/variance_ratio.R

```diff
@@ -33,100 +33,162 @@
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          method = variance_ratio_matrixdata,
+                          bootnumber = bootnumber)
+
+      ## confidence intervals
+      null_output <- lapply(null_list, `[[`, i = "out")
+
+      null_means <- vector(mode = "numeric", length = bootnumber)
+
+      for (B in 1:bootnumber) {
+        null_means[B] <- mean(vapply(null_output, `[[`, 1, i = B))
+      }
+
+      li <- (1 - level)/2
+      ui <- 1 - li
+      lowerCI <- quantile(null_means, li)
+      upperCI <- quantile(null_means, ui)
+      nullmean <- mean(null_means)
+      nullout <- data.frame(lowerCI = lowerCI,
+                            upperCI = upperCI,
+                            nullmean =  nullmean)
+
+      # bind actual value and CI
+      output <- cbind(nullout, VR)
+
     } else {
-        
-      # if multiple replicates, check all replicates have values
-        check_single(df, time.var, species.var, replicate.var)
-        df[replicate.var] <- if(is.factor(df[[replicate.var]]) == TRUE) {
-            factor(df[[replicate.var]])
-        } else {
-            df[replicate.var]
-        }
-        
-        # calculate average variance ratio across replicates
-        if(average.replicates == TRUE) {
-           check_multispp(df, species.var, replicate.var)
-            df <- df[order(df[[replicate.var]]),]
-            X <- split(df, df[replicate.var])
-            VR <- mean(unlist(lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var))) 
-        } else {
-          
-          # calculate the variance ratio for replicate
-            check_multispp(df, species.var, replicate.var)
-            df <- df[order(df[[replicate.var]]),]
-            X <- split(df, df[replicate.var])
-            VR <- do.call("rbind", lapply(X, FUN = variance_ratio_longformdata, time.var, species.var, abundance.var))
-        }
+
+      # calculate the variance ratio for replicate
+      check_multispp(df, species.var, replicate.var)
+      df <- df[order(df[[replicate.var]]),]
+      X <- split(df, df[replicate.var])
+
+      ## also do separate null models (ie on split list)
+      null_list <- lapply(X, FUN = cyclic_shift,  time.var = time.var,
+                          species.var = species.var,
+                          abundance.var = abundance.var,
+                          method = variance_ratio_matrixdata,
+                          bootnumber = bootnumber)
+      ## calculate confidence intervals for each:
+      null_intervals <- lapply(null_list, confint)
+
+      ## combine as data.frames and preserve replicate var
+      nullout <- purrr::map_df(null_intervals, as.data.frame, .id = replicate.var)
+
+      VR_list <- lapply(X, FUN = variance_ratio_longformdata,
+                        time.var, species.var, abundance.var)
+
+      VRout <- purrr::map_df(VR_list, ~data.frame(VR = .), .id = replicate.var)
```

Comment:
@laurenmh @aammd purrr has dependencies on a number of other large packages, including magrittr (>= 1.5), dplyr (>= 0.4.3), Rcpp, lazyeval.  So this definitely increases the overhead of installing codyn.  That said, these are pretty standard packages these days, so probably not a big deal for most people.  Each time a package gets updated in some incompatible way, we will have to release a new version of codyn that fixes the issue. This happens more than you might think (most recently with Hadley's `testthat`. So just be prepared for more maintenance time with each dependency we add.